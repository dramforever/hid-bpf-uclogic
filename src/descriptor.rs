use std::{collections::HashMap, error::Error, fmt::Display};

use minijinja;
use serde::Serialize;

#[derive(Debug)]
pub(crate) struct DeviceInfo {
    pub firmware: String,
    pub magic_bytes: Vec<u8>,
}

#[derive(Debug, Clone, Copy, Serialize)]
pub(crate) struct ParsedDeviceInfo {
    pub x_max: u32,
    pub y_max: u32,
    pub pres_max: u16,
    pub resolution: u16,
    pub num_btns: u8,
}

impl DeviceInfo {
    pub(crate) fn from_str(text: &str) -> Result<Self, Box<dyn Error>> {
        fn unquote(s: &str) -> Result<&str, &'static str> {
            s.strip_prefix('"')
                .and_then(|s| s.strip_suffix('"'))
                .ok_or("Incorrect quotes")
        }

        let info: HashMap<&str, &str> = text
            .lines()
            .filter_map(|l| l.split_once('='))
            .map(|(key, value)| Ok((key, unquote(value)?)))
            .collect::<Result<_, &'static str>>()?;

        if info.contains_key("HUION_PAD_MODE") {
            return Err("Unsupported v1 protocol device")?;
        }

        Ok(Self {
            firmware: info
                .get("HUION_FIRMWARE_ID")
                .copied()
                .ok_or("No HUION_FIRMWARE_ID found")?
                .to_owned(),
            magic_bytes: hex::decode(
                info.get("HUION_MAGIC_BYTES")
                    .copied()
                    .ok_or("No HUION_MAGIC_BYTES found")?,
            )?,
        })
    }

    pub(crate) fn parse(&self) -> Result<ParsedDeviceInfo, &'static str> {
        if self.magic_bytes.len() < 18 {
            return Err("Device info too short");
        }

        if self.magic_bytes[0] as usize != self.magic_bytes.len() {
            return Err("Device info has incorrect length");
        }

        fn le(bytes: &[u8]) -> u32 {
            debug_assert!(bytes.len() <= 4);
            bytes
                .iter()
                .copied()
                .rev()
                .fold(0, |acc, x| (acc << 8) | x as u32)
        }

        let m = &self.magic_bytes;

        Ok(ParsedDeviceInfo {
            x_max: le(&m[2..][..3]),
            y_max: le(&m[5..][..3]),
            pres_max: le(&m[8..][..2]) as _,
            resolution: le(&m[10..][..2]) as _,
            num_btns: m[13] as _,
        })
    }
}

impl ParsedDeviceInfo {
    pub(crate) fn descriptor(&self) -> Result<Vec<u8>, Box<dyn Error>> {
        let mut env = minijinja::Environment::new();
        fn bytes(bs: &[u8]) -> String {
            bs.iter()
                .map(|b| format!("{b:02x}"))
                .collect::<Vec<_>>()
                .join(" ")
        }
        env.add_filter("u16", |val: u16| bytes(&val.to_le_bytes()));
        env.add_filter("u32", |val: u32| bytes(&val.to_le_bytes()));
        env.add_filter("u8", |val: u8| format!("{val:02x}"));
        let mut hex_str = env.render_str(include_str!("descriptor.j2"), self)?;
        hex_str.retain(|x| !x.is_ascii_whitespace());
        Ok(hex::decode(hex_str).expect("Invalid hex data"))
    }
}

impl Display for ParsedDeviceInfo {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        let x_physical = self.x_max as f64 / self.resolution as f64;
        let y_physical = self.y_max as f64 / self.resolution as f64;

        write!(
            f,
            "Device with {} buttons, max pen pressure {}, logical size ({}, {}), resolution {}, physical size in inches ({:.2}, {:.2})",
            self.num_btns,
            self.pres_max,
            self.x_max,
            self.y_max,
            self.resolution,
            x_physical,
            y_physical,
        )
    }
}
