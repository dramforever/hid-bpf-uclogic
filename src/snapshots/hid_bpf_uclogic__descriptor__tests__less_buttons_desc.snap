---
source: src/descriptor.rs
expression: descriptor_dump(&desc)
---
R: 275 05 0d 09 01 a1 01 85 08 09 20 a1 00 15 00 25 01 75 01 09 42 09 44 09 5a 95 03 81 02 95 04 81 01 09 32 95 01 81 02 75 18 95 01 a4 05 01 65 13 55 0d 17 00 00 00 00 27 70 c6 00 00 37 00 00 00 00 47 10 27 00 00 09 30 81 02 17 00 00 00 00 27 06 7c 00 00 37 00 00 00 00 47 6a 18 00 00 09 31 81 02 b4 16 00 00 26 ff 1f 09 30 75 10 95 01 81 02 75 08 95 02 a4 65 14 55 00 15 c4 35 c4 25 3c 45 3c 09 3d 09 3e 81 02 b4 c0 c0 05 01 09 07 a1 01 85 03 15 00 25 01 05 0d 09 39 a1 00 09 44 95 01 75 01 81 02 95 07 81 01 05 01 09 30 09 31 95 02 75 08 81 02 c0 05 09 19 01 29 05 95 05 75 01 81 02 95 3b 81 01 c0 05 01 09 07 a1 01 85 f0 15 00 25 01 05 0d 09 39 a1 00 05 09 09 01 95 01 75 01 81 02 95 07 81 01 05 01 09 30 09 31 75 08 95 02 81 02 95 01 81 01 05 01 09 38 95 01 15 ff 25 01 81 06 05 0c 0a 38 02 95 01 81 06 c0 75 08 95 05 81 01 c0
